### PR TITLE
refactor(sink): refine sink trait and macro

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

In this PR we refine sink trait and macro similar to what we do for source in https://github.com/risingwavelabs/risingwave/pull/12260.

Previously when we add a new sink, we need to add new code to `SinkConfig`, `SinkImpl`, `build_sink`, `dispatch_sink_impl` and other places. After this PR, we have a macro `for_all_sinks` that declare all sinks, and when we add a new sink, we only need to add the sink to the list in `for_all_sinks`. The `SinkConfig` enum is removed, and the `dispatch_sink_impl` will generate code based on the sinks listed in `for_all_sinks`.

Sinks will implement `TryFrom<SinkParam>` as a general way to be created from the sink param.

Besides, the `async_trait` macro on sink trait and impl are removed.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/sink/mod.rs [hunk 1]
```diff
@@ -152,13 +206,14 @@ pub struct SinkWriterParam {
     pub meta_client: Option<MetaClient>,
 }
 
-#[async_trait]
-pub trait Sink {
+pub trait Sink: TryFrom<SinkParam, Error = SinkError> {
```
Comment (liurenjie1024) [inline src/connector/src/sink/mod.rs:209]:
```suggestion
pub trait Sink: TryFrom<SinkContext, Error = SinkError> {
```

I would suggest to add sth like `SinkContext`, we may need to add more info such as actor id for better observability.

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM

**Review by StrikeW — APPROVED** [review]
LGTM

**Review by yufansong — APPROVED** [review]
Nats part LGTM!

File: src/connector/src/sink/mod.rs (hunk 1 above)
Comment (wenym1) [reply src/connector/src/sink/mod.rs:212]: The `Sink` is a set of logical metadata, and it is used in all env such an CN and meta. The compute runtime information, such as actor id and executor id, should only be passed to the sink writer, which is already passed in the `SinkWriterParam` when we create a new sink writer.

**Pull request merged**